File: Algoritmos/02-AddTwoNumbers.cs
using System;

namespace AddTwoNumbers
{
	public class LinkedList
	{
        private ListNode head;
        private ListNode current;

        public LinkedList()
        {
            head = new ListNode();
            head.next = null;
        }
        public class ListNode
		{
			public int val;
			public ListNode next;
			public ListNode(int val = 0, ListNode next = null)
			{
				this.val = val;
				this.next = next;
			}
		}

        public void AddNode(int pDato)
        {
            current = head;
            //Recorrer hasta encontrar el final de la lista

            while (current.next != null)
            {
                //Avanzar al siguiente nodo
                current = current.next;
            }
            //Crear un nuevo nodo
            ListNode temp = new ListNode();

            //Insertamos el dato
            temp.val = pDato;

            //Finalizamos correctamente
            temp.next = null;

            //Insertamos el nuevo nodo al final de la lista, ligar el último nodo encontrado con el nuevo nodo
            current.next = temp;
        }

        public void Transverse()
        {
            //Trabajo al inicio
            current= head;

            //Recorrer hasta encontrar el final de la lista
            while (current.next != null)
            {
                //Avanzar al siguiente nodo
                current = current.next;
                //Obtener el siguiente nodo y lo mostramos
                int d = current.val;

                Console.Write("[{0}] ", d);
            }
            Console.WriteLine();
        }
		public ListNode AddTwoNumbers(ListNode l1, ListNode l2) 
    	{
            var dummy = new ListNode(-1);
            var current = dummy;

            var carry = 0;
            while (l1 != null || l2 != null)
            {
                var value1 = l1 == null ? 0 : l1.val;
                var value2 = l2 == null ? 0 : l2.val;

                var sum = value1 + value2 + carry;
                carry = sum / 10;
                sum %= 10;
                current.next = new ListNode(sum);

                current = current.next;
                l1 = l1 == null ? null : l1.next;
                l2 = l2 == null ? null : l2.next;
            }

            if (carry != 0)
                current.next = new ListNode(carry);

            return dummy.next;
        }

		public static void Main(string[] args)
		{
			//Initialize linked list
            LinkedList list1 = new LinkedList();
            LinkedList list2 = new LinkedList();

            //Add nodes to linked list
            list1.AddNode(2);
            list1.AddNode(4);
            list1.AddNode(3);

            list2.AddNode(5);
            list2.AddNode(6);
            list2.AddNode(4);

            //Print linked list
            Console.Write("Input: l1 = "); 
            list1.Transverse();
            Console.Write("Input: l2 = ");
            list2.Transverse();

            Console.WriteLine();

            //Add two numbers
            LinkedList list3 = new LinkedList();

            var result = list3.AddTwoNumbers(list1.head, list2.head);
            list3.head = result;

            //Print linked list
            Console.Write("Output: l3 = ");
            list3.Transverse();

        
        }
	}
}

File: Algoritmos/02-AddTwoNumbers.cpp
#include <iostream>

struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* next) : val(x), next(next) {}
};

ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
    int carry = 0;
    ListNode dummy, *current = &dummy;
    while (l1 || l2 || carry) {
        if (l1) {
            carry += l1->val;
            l1 = l1->next;
        }
        if (l2) {
            carry += l2->val;
            l2 = l2->next;
        }
        //Insert a new node with the value of carry % 10
        current->next = new ListNode(carry % 10);

        //Move current to the next node and update carry
        current = current->next;
        carry /= 10;
    }

    //Return the next node of the dummy node, which is the head of the new list
    return dummy.next;
}

void printList(ListNode* n)
{
    while (n != NULL)
    {
        std::cout << "[" << n->val << "] -> ";
        n = n->next;
    }

    std::cout << "NULL";
}

int main()
{
    //Build the first linked list
    ListNode* l1 = new ListNode();
    ListNode* l2 = new ListNode();
    ListNode* l3 = new ListNode();

    l1->val = 2;
    l1->next = l2;

    l2->val = 4;
    l2->next = l3;

    l3->val = 3;
    l3->next = NULL;

    std::cout << "Input: l1 =  " ;
    printList(l1);
    std::cout << ", ";

    //Build the second linked list
    ListNode* l4 = new ListNode();
    ListNode* l5 = new ListNode();
    ListNode* l6 = new ListNode();

    l4->val = 5;
    l4->next = l5;
    l5->val = 6;
    l5->next = l6;
    l6->val = 4;
    l6->next = NULL;

    std::cout << "l2 =  " ;
    printList(l4);
    std::cout << "\n";

    //Add the two linked lists
    ListNode* l7 = addTwoNumbers(l1, l4);

    std::cout << "Output: ";
    printList(l7);
    std::cout << "\n";

    //std::cin.get();

}
